include/crm/services.h

```diff
@@ -292,8 +292,10 @@ gboolean services_action_kick(const char *name, const char *action,
  * \param[in] action_callback       Function to call when the action completes
  * \param[in] action_fork_callback  Function to call after action process forks
  *
- * \return TRUE if execution was successfully initiated, FALSE otherwise (in
- *              which case the callback will not be called)
+ * \return TRUE if the action should not be freed by the caller (e.g., if
+ *         execution was initiated successfully, if execution failed to initiate
+ *         and the action has already been freed, or if the action is blocked);
+ *         FALSE otherwise
```

Comment:
Ha ha, this is where I started, too. After tracing through all the possibilities, I realized it's completely messed up. This is what I have at the moment:

    + * \return Boolean value
    + *
    + * \retval TRUE if the caller should not free or otherwise use \p op again,
    + *         because one of these conditions is true:
    + *
    + *         * \p op is NULL.
    + *         * The action was successfully initiated, in which case
    + *           \p action_fork_callback has been called, but \p action_callback has
    + *           not (it will be called when the action completes).
    + *         * The action's ID matched an existing recurring action. The existing
    + *           action has taken over the callback and callback data from \p op
    + *           and has been re-initiated asynchronously, and \p op has been freed.
    + *         * Another action for the same resource is in flight, and \p op will
    + *           be blocked until it completes.
    + *         * The action could not be initiated, and is either non-recurring or
    + *           being cancelled. \p action_fork_callback has not been called, but
    + *           \p action_callback has, and \p op has been freed.
    + *
    + * \retval FALSE if \op is still valid, because the action cannot be initiated,
    + *         and is a recurring action that is not being cancelled.
    + *         \p action_fork_callback has not been called, but \p action_callback
    + *         has, and a timer has been set for the next invocation of \p op.

I was surprised to find that only a single code path leads to FALSE, and the callback *is* called there, which makes the return value irrelevant (the callers' callbacks free the user data, so the user data can never be assumed to be valid after calling this function). I would be grateful if you could verify that my description of the possible paths is accurate.

I hadn't gotten as far as a fix, but I was thinking we should verify that the action_callback is *always* called (either in this function, or on async completion), and FALSE should be returned solely in the case that the svc_action_t is complete and can be freed (with the flip side being that the svc_action_t and its used data should never be used again by the caller if the return value is TRUE).

BTW that first condition (op is NULL) is a CRM_CHECK() I was planning to add.

A related issue I found is that services_action_async_fork_notify() -> handle_duplicate_recurring() -> recurring_action_timer() calls services_action_async() but ignores the return value, meaning services_action_async_fork_notify() will always return TRUE in that case, even if services_action_async() returned FALSE. We'll need to get the correct value back to the original caller. The guts should be a new function used in most places, and recurring_timer() should be a wrapper for it used only with g_timeout_add(). I hadn't figured out yet what should happen if services_action_async() returns FALSE in the timer.

The callers need to be updated to handle the memory properly. lrmd_rsc_execute_service_lib() currently uses the user data if the function returns FALSE, which is bad, and presumably hasn't caused problems only because the FALSE condition is rare and the memory probably still has the same contents at that point despite being freed.

Reply:
I started to do an exhaustive doc comment and was afraid it would be too verbose, so I tried to summarize (and omitted the duplicate recurring op case). I like your enumerative one.

> I was surprised to find that only a single code path leads to FALSE, and the callback is called there, which makes the return value somewhat irrelevant (the callers' callbacks free the user data, so the user data can never be assumed to be valid after calling this function).

Somewhat, yes.

- If it returns TRUE, then we know it's unsafe to free or use the op. The user data and/or the svc_action may have already freed.
- If it returns FALSE, things are a tad different. It's unsafe to use the user data, becuase the done callback has been run, which frees it. The svc_action_t has **not** been freed, so we need to take care of that in the caller.

That seems to line up what you want to verify in your next paragraph.

The impact on `internal_stonith_action_execute()` is that the last call to `set_result_from_svc_action()` will segfault if `async_fork_notify()` returns FALSE. It uses the user data, which we know has been freed. We need to skip that call for async actions. We'd still need to keep the `svc_action->params = NULL` and `services_action_free(svc_action)` lines.

> I would be grateful if you could verify that my description of the possible paths is accurate.

It looks like it. Here's a stripped-down pseudocode view. The only way I see to get FALSE is if a non-duplicate, non-blocked recurring op fails to initiate and then gets rescheduled instead of canceled.

```
services_action_async_fork_notify(op):
  if recurring:
    if duplicate exists:	// handle_duplicate_recurring(op)
      update dup's user data and reschedule dup
      free op
      return TRUE
    add to recurring actions

  if blocked:
    return TRUE

  return execute_action(op) == pcmk_rc_ok
  
execute_action(op):
  if successfully initiated:
    if not upstart and not systemd:
      call fork_callback
    return pcmk_rc_ok
  return services__finalize_async_op(op)
  
services__finalize_async_op(op):
  if recurring:
    if op->cancel:	// see note 1 below
      free (op->exit_reason)
      remove from recurring_actions	// cancel_recurring_action(op)
    else:
      reschedule op

  run done callback if it exists
  if recurring and not op->cancel:
    return EBUSY
  free op
  return pcmk_rc_ok
```

[1] I see only one location where op->cancel is set to TRUE: in services_action_cancel(). cancel_recurring_action() is also called there, so we log a duplicate message and make two attempts to remove from the hashtable.
[2] Hopefully the services_action_cleanup(op) that's in services_action_cancel() for systemd/upstart can never run for the same op as finalize_async runs.

---

Is it okay that we run the done callback for a recurring op that's rescheduled instead of canceled -- definitely won't need the user data later?

---

In all likelihood I'll have to drop active work on the PR for the time being (other than discussion), so feel free to pick it back up. I was hoping to find a fairly straightforward fix (achieved for the test case, but there are lingering issues).

P.S. I missed updating the doc comments for `services__execute_upstart()` and `services__execute_systemd()` when I hit the others. They should stay in alignment with regard to retvals.